File: android/src/main/java/com/xpel/flutter_lightweight_store/FlutterLightweightStorePlugin.java
package com.xpel.flutter_lightweight_store;

import androidx.annotation.NonNull;

import com.tesla.modules.store.TlStoreManager;
import com.tesla.modules.store.util.TlStoreLogger;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * FlutterLightweightStorePlugin
 */
public class FlutterLightweightStorePlugin implements FlutterPlugin, MethodCallHandler {
    private MethodChannel channel;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_lightweight_store");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        String method = call.method;
        Object arguments = call.arguments;
        TlStoreLogger.log("[FlutterLightweightStorePlugin] receive: " + method + ", arguments: " + arguments);
        // 约定一定要传 Map 类型参数, 空也不行
        if (!(arguments instanceof Map)) {
            _sendError(result, 3000300, null, "arguments type should be a hash map.");
            return;
        }
        try {
            handleMethod(method, (Map) arguments, result);
        } catch (Exception e) {
            _sendError(result, 3000301, null, "exception: " + e.getMessage());
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    private static void _sendError(@NonNull Result result, int code, String msg, String error) {
        Map<String, Object> resultMap = new HashMap<>();
        resultMap.put("code", code);
        if (msg != null) {
            resultMap.put("msg", msg);
        }
        if (error != null) {
            resultMap.put("error", error);
        }
        _send(result, resultMap);
    }

    private static void _sendSuccess(@NonNull Result result, Object data) {
        Map<String, Object> resultMap = new HashMap<>();
        resultMap.put("code", 200);
        if (data != null) {
            resultMap.put("data", data);
        }
        _send(result, resultMap);
    }

    private static void _send(@NonNull Result result, @NonNull Map map) {
        TlStoreLogger.log("[FlutterLightweightStorePlugin] send: " + map);
        result.success(map);
    }


    private void handleMethod(@NonNull String method, @NonNull Map arguments, @NonNull Result result) {
        JSONObject args = new JSONObject(arguments);
        String module = args.optString("module");
        if (module.isEmpty()) {
            _sendError(result, 3000302, null, "arguments should have a 'module' KV.");
            return;
        }
        if ("register".equals(method)) {
            String aesKey = args.optString("aesKey");
            String aesIV = args.optString("aesIV");
            Boolean isAsyncApply = args.optBoolean("isAsyncApply");
            Boolean isKeepKeyClearText = args.optBoolean("isKeepKeyClearText");
            TlStoreManager.getInstance().register(module, aesKey.isEmpty() ? null : aesKey, aesIV.isEmpty() ? null : aesIV, isAsyncApply, isKeepKeyClearText);
            _sendSuccess(result, null);
            return;
        }
        if ("unregister".equals(method)) {
            TlStoreManager.getInstance().unregister(module);
            _sendSuccess(result, null);
            return;
        }
        String key = args.optString("key");
        if (key.isEmpty()) {
            _sendError(result, 3000303, null, "arguments should have a 'key' KV in get* method.");
            return;
        }
        if ("contains".equals(method)) {
            boolean retValue = TlStoreManager.getInstance().contains(module, key);
            _sendSuccess(result, retValue);
            return;
        }
        if ("removeKey".equals(method)) {
            TlStoreManager.getInstance().removeKey(module, key);
            _sendSuccess(result, true); // true always now ...
            return;
        }
        if ("getString".equals(method)) {
            String retValue = TlStoreManager.getInstance().getString(module, key, null);
            _sendSuccess(result, retValue);
            return;
        }
        if ("getInt".equals(method) || "getLong".equals(method)) {
            long retValue = TlStoreManager.getInstance().getLong(module, key, 0);
            boolean isKeyExisted = true;
            if (retValue == 0) {
                isKeyExisted = TlStoreManager.getInstance().contains(module, key);
            }
            _sendSuccess(result, isKeyExisted ? retValue : null);
            return;
        }
        if ("getFloat".equals(method) || "getDouble".equals(method)) {
            double retValue = TlStoreManager.getInstance().getDouble(module, key, 0.0);
            boolean isKeyExisted = true;
            if (retValue == 0.0) {
                isKeyExisted = TlStoreManager.getInstance().contains(module, key);
            }
            _sendSuccess(result, isKeyExisted ? retValue : null);
            return;
        }
        if ("getBoolean".equals(method)) {
            boolean retValue = TlStoreManager.getInstance().getBoolean(module, key, false);
            boolean isKeyExisted = true;
            if (!retValue) {
                isKeyExisted = TlStoreManager.getInstance().contains(module, key);
            }
            _sendSuccess(result, isKeyExisted ? retValue : null);
            return;
        }

        if (!args.has("value")) {
            _sendError(result, 3000304, null, "arguments should have a 'value' KV in set* method.");
            return;
        }
        if ("setString".equals(method)) {
            TlStoreManager.getInstance().setString(module, key, args.optString("value"));
            _sendSuccess(result, null);
            return;
        }
        if ("setInt".equals(method) || "setLong".equals(method)) {
            TlStoreManager.getInstance().setLong(module, key, args.optLong("value"));
            _sendSuccess(result, null);
            return;
        }
        if ("setFloat".equals(method) || "setDouble".equals(method)) {
            TlStoreManager.getInstance().setDouble(module, key, args.optDouble("value"));
            _sendSuccess(result, null);
            return;
        }
        if ("setBoolean".equals(method)) {
            TlStoreManager.getInstance().setBoolean(module, key, args.optBoolean("value"));
            _sendSuccess(result, null);
            return;
        }
        result.notImplemented();
    }
}

File: ios/Classes/FlutterLightweightStorePlugin.m
#import "FlutterLightweightStorePlugin.h"
#import "TlStoreManager.h"

#import "EncryptorOfAES.h"

@implementation FlutterLightweightStorePlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"flutter_lightweight_store" binaryMessenger:[registrar messenger]];
    FlutterLightweightStorePlugin* instance = [[FlutterLightweightStorePlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}


- (void)resultFailed:(FlutterResult)result errorCode:(int)errorCode errorMsg:(NSString *)errorMsg {
    result(@{@"code": @(errorCode), @"error": errorMsg ?: @""});
}

#define kFailedResult(_result, _code, _msg) [self resultFailed:_result errorCode:_code errorMsg:_msg]

- (void)resultSuccess:(FlutterResult)result data:(nullable id)data {
    if (result == nil) return;
    if (data == nil) {
        result(@{@"code": @(200)});
    } else {
        result(@{@"code": @(200), @"data": data});
    }
}

#define kSuccessResult(_result, _data) [self resultSuccess:_result data:_data]


# pragma mark - Override Methods
- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString *method = call.method;
    id arguments = call.arguments;
    if (![arguments isKindOfClass:NSDictionary.class]) {
        kFailedResult(result, 3000300, @"arguments type should be a hash map.");
        return;
    }
    NSDictionary *args = arguments;
    
    @try {
        [self handleMethod:method arguments:args result:result];
    } @catch (NSException *exception) {
        kFailedResult(result, 3000301, [exception description]);
    } @finally {
        // nothing now ...
    }
}

- (void)handleMethod:(NSString * _Nonnull)method arguments:(NSDictionary * _Nonnull)arguments result:(FlutterResult)result {
    NSString *module = [arguments objectForKey:@"module"];
    if (module == nil || module.length == 0) {
        kFailedResult(result, 3000302, @"arguments should have a 'module' KV.");
        return;
    }
    
    if ([method isEqualToString:@"register"]) {
        NSString *aesKey = [arguments objectForKey:@"aesKey"];
        NSString *aesIV = [arguments objectForKey:@"aesIV"];
        BOOL isKeepKeyClearText = [[arguments objectForKey:@"isKeepKeyClearText"] boolValue];
        
        [TlStoreManager.instance registration:module aesKey:aesKey aesIV:aesIV isKeepKeyClearText:isKeepKeyClearText];
        kSuccessResult(result, nil);
        return;
    }
    
    if ([method isEqualToString:@"unregister"]) {
        [TlStoreManager.instance unregistration:module];
        kSuccessResult(result, nil);
        return;
    }
    
    NSString *key = [arguments objectForKey:@"key"];
    if (key == nil || key.length == 0) {
        kFailedResult(result, 3000303, @"arguments should have a 'key' KV in get* method.");
        return;
    }
    
    if ([method isEqualToString:@"contains"]) {
        BOOL retValue = [TlStoreManager.instance contains:module forKey:key];
        kSuccessResult(result, @(retValue));
        return;
    }
    
    if ([method isEqualToString:@"removeKey"]) {
        [TlStoreManager.instance removeKey:module forKey:key];
        BOOL retValue = YES;
        kSuccessResult(result, @(retValue));
        return;
    }
    
    // getter
    if ([method isEqualToString:@"getString"]) {
        NSString *retValue = [TlStoreManager.instance getString:module forKey:key];
        kSuccessResult(result, retValue);
        return;
    }
    
    if ([method isEqualToString:@"getInt"] || [method isEqualToString:@"getLong"]) {
        long retValue = [TlStoreManager.instance getLong:module forKey:key];
        BOOL isKeyExisted = YES;
        if (retValue == 0) {
            isKeyExisted = [TlStoreManager.instance contains:module forKey:key];
        }
        kSuccessResult(result, isKeyExisted ? @(retValue) : nil);
        return;
    }
    
    if ([method isEqualToString:@"getFloat"] || [method isEqualToString:@"getDouble"]) {
        double retValue = [TlStoreManager.instance getDouble:module forKey:key];
        BOOL isKeyExisted = YES;
        if (retValue == 0.0) {
            isKeyExisted = [TlStoreManager.instance contains:module forKey:key];
        }
        kSuccessResult(result, isKeyExisted ? @(retValue) : nil);
        return;
    }
    
    if ([method isEqualToString:@"getBoolean"]) {
        BOOL retValue = [TlStoreManager.instance getBoolean:module forKey:key];
        BOOL isKeyExisted = YES;
        if (retValue == FALSE) {
            isKeyExisted = [TlStoreManager.instance contains:module forKey:key];
        }
        kSuccessResult(result, isKeyExisted ? @(retValue) : nil);
        return;
    }
    
    
    NSObject *value = [arguments objectForKey:@"value"];
    if (value == nil) {
        kFailedResult(result, 3000304, @"arguments should have a 'value' KV in set* method.");
        return;
    }
    
    // setter
    if ([method isEqualToString:@"setString"]) {
        NSString *setVal = nil;
        if ([value isKindOfClass:NSString.class]) {
            setVal = (NSString *) value;
        } else if ([value isKindOfClass:NSNumber.class]) {
            setVal = [(NSNumber *)value stringValue];
        } else {
            setVal = [value description];
        }
        [TlStoreManager.instance setString:module forKey:key value:setVal];
        kSuccessResult(result, nil);
        return;
    }
    
    if ([method isEqualToString:@"setInt"] || [method isEqualToString:@"setLong"]) {
        long setVal = 0;
        if ([value isKindOfClass:NSString.class]) {
            setVal = [(NSString *)value integerValue];
        } else if ([value isKindOfClass:NSNumber.class]) {
            setVal = [(NSNumber *)value integerValue];
        } else {
            kFailedResult(result, -1, @"value not a int/long value");
            return;
        }
        [TlStoreManager.instance setLong:module forKey:key value:setVal];
        kSuccessResult(result, nil);
        return;
    }
    
    if ([method isEqualToString:@"setFloat"] || [method isEqualToString:@"setDouble"]) {
        double setVal = 0;
        if ([value isKindOfClass:NSString.class]) {
            setVal = [(NSString *)value doubleValue];
        } else if ([value isKindOfClass:NSNumber.class]) {
            setVal = [(NSNumber *)value doubleValue];
        } else {
            kFailedResult(result, -1, @"value not a float/double value");
            return;
        }
        [TlStoreManager.instance setDouble:module forKey:key value:setVal];
        kSuccessResult(result, nil);
        return;
    }
    
    if ([method isEqualToString:@"setBoolean"]) {
        BOOL setVal = 0;
        if ([value isKindOfClass:NSString.class]) {
            setVal = [(NSString *)value boolValue];
        } else if ([value isKindOfClass:NSNumber.class]) {
            setVal = [(NSNumber *)value boolValue];
        } else {
            kFailedResult(result, -1, @"value not a boolean value");
            return;
        }
        [TlStoreManager.instance setBoolean:module forKey:key value:setVal];
        kSuccessResult(result, nil);
        return;
    }
    
    // TODO ... Remove the following code after several versions
    if ([method isEqualToString:@"restoreEmptyAesKeyIvContents"]) {
        NSUserDefaults *user = [[NSUserDefaults alloc] initWithSuiteName:module];
        if ([user objectForKey:@"__already_restored__"] == nil) {
            [user setObject:@(YES) forKey:@"__already_restored__"];

            NSDictionary<NSString *, id> *allKeysValues = [user dictionaryRepresentation];
            NSArray<NSString *> *allKeys = allKeysValues.allKeys;

            for (NSString *key in allKeys) {
                if ([key hasPrefix:@"Apple"]) continue;
                id value = [allKeysValues valueForKey:key];
                if (![value isKindOfClass:NSString.class]) continue;

                NSString *deKey = [self emptyAesKeyIvDecrypt:key];
                NSString *deValue = [self emptyAesKeyIvDecrypt:value];
                NSLog(@"Key: %@, Value: %@. De-key: %@, De-value: %@", key, value, deKey, deValue);
                if (deKey == nil || deValue == nil) continue;
                [user setObject:deValue forKey:deKey];
            }
        }

        kSuccessResult(result, nil);
        return;
    }

    result(FlutterMethodNotImplemented);
}

- (NSString *)emptyAesKeyIvEncrypt: (NSString *)content {
    return [EncryptorOfAES encryptToBase64:[content dataUsingEncoding:NSUTF8StringEncoding] Key:@"" IV:@""];
}

- (NSString *)emptyAesKeyIvDecrypt: (NSString *)content {
    return [[NSString alloc] initWithData:[EncryptorOfAES decryptFromBase64:content Key:@"" IV:@""] encoding:NSUTF8StringEncoding];
}

@end
